### PR TITLE
[Improvement] Only report to the shuffle servers that owns the blocks

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

Only report to the shuffle servers that owns the blocks

### Why are the changes needed?
I found some unnecessary log is shown in the shuffle server's log, like this 

```
[INFO] 2023-02-02 10:59:04,045 Grpc-881 ShuffleServerGrpcService reportShuffleResult - Report 0 blocks as shuffle result for the task of appId[application_1672821343673_3878453_1675303644304], shuffleId[1619], taskAttemptId[1198007]
```

As shown above, there is no necessary to report 0 blocks to shuffle server for client. Let's remove this logic.

__Improvement__
1.  for shuffle server, it reduces the log size and make its more clear
2. for client, it speeds up the client reporting due to avoiding unnecessary rpc request

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
1. UTs

### CODE REVIEW

File: client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java [hunk 1]
```diff
@@ -520,18 +520,28 @@ public void reportShuffleResult(
...
+      int partitionIdx = entry.getKey();
       for (ShuffleServerInfo ssi : entry.getValue()) {
         if (!groupedPartitions.containsKey(ssi)) {
-          groupedPartitions.putIfAbsent(ssi, Lists.newArrayList());
+          groupedPartitions.put(ssi, Lists.newArrayList());
         }
-        groupedPartitions.get(ssi).add(entry.getKey());
+        groupedPartitions.get(ssi).add(partitionIdx);
+      }
+      if (CollectionUtils.isNotEmpty(partitionToBlockIds.getOrDefault(partitionIdx, null))) {
```
Comment (xianjingfeng) [inline client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java:530]: Is `partitionToBlockIds.getOrDefault(partitionIdx, null)`  same as `partitionToBlockIds.get(partitionIdx)`?

File: integration-test/common/src/test/java/org/apache/uniffle/test/ShuffleWithRssClientTest.java [hunk 2]
```diff
@@ -150,6 +150,57 @@ public void rpcFailTest() throws Exception {
...
+        testAppId,
+        1,
+        Lists.newArrayList(new PartitionRange(2, 2)),
+        new RemoteStorageInfo(""),
+        ShuffleDataDistributionType.NORMAL
+    );
+
+    Map<Integer, List<ShuffleServerInfo>> partitionToServers = Maps.newHashMap();
+    partitionToServers.putIfAbsent(1, Lists.newArrayList(shuffleServerInfo1));
+    partitionToServers.putIfAbsent(2, Lists.newArrayList(shuffleServerInfo2));
```
[the remark this answers is not in the document]
Comment (zuston) [reply integration-test/common/src/test/java/org/apache/uniffle/test/ShuffleWithRssClientTest.java:177]: Nice catch.

[the remark this answers is not in the document]
Comment (zuston) [reply integration-test/common/src/test/java/org/apache/uniffle/test/ShuffleWithRssClientTest.java]: Done

**Review by xianjingfeng — DISMISSED** [review]
LGTM

File: client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java [hunk 3]
```diff
@@ -520,18 +520,28 @@ public void reportShuffleResult(
     Map<ShuffleServerInfo, List<Integer>> groupedPartitions = Maps.newHashMap();
     Map<Integer, Integer> partitionReportTracker = Maps.newHashMap();
     for (Map.Entry<Integer, List<ShuffleServerInfo>> entry : partitionToServers.entrySet()) {
+      int partitionIdx = entry.getKey();
       for (ShuffleServerInfo ssi : entry.getValue()) {
         if (!groupedPartitions.containsKey(ssi)) {
-          groupedPartitions.putIfAbsent(ssi, Lists.newArrayList());
+          groupedPartitions.put(ssi, Lists.newArrayList());
```
[the remark this answers is not in the document]
Comment (zuston) [reply client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java]: Yes. It's also OK.

File: client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java [hunk 4]
```diff
@@ -520,18 +520,28 @@ public void reportShuffleResult(
...
+        partitionReportTracker.putIfAbsent(partitionIdx, 0);
       }
-      partitionReportTracker.putIfAbsent(entry.getKey(), 0);
     }
+
     for (Map.Entry<ShuffleServerInfo, List<Integer>> entry : groupedPartitions.entrySet()) {
       Map<Integer, List<Long>> requestBlockIds = Maps.newHashMap();
       for (Integer partitionId : entry.getValue()) {
-        requestBlockIds.put(partitionId, partitionToBlockIds.get(partitionId));
+        List<Long> partitions = partitionToBlockIds.get(partitionId);
```
[the remark this answers is not in the document]
Comment (zuston) [reply client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java]: Done

File: client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java [hunk 5]
```diff
@@ -520,18 +520,26 @@ public void reportShuffleResult(
     Map<ShuffleServerInfo, List<Integer>> groupedPartitions = Maps.newHashMap();
     Map<Integer, Integer> partitionReportTracker = Maps.newHashMap();
     for (Map.Entry<Integer, List<ShuffleServerInfo>> entry : partitionToServers.entrySet()) {
+      int partitionIdx = entry.getKey();
       for (ShuffleServerInfo ssi : entry.getValue()) {
-        if (!groupedPartitions.containsKey(ssi)) {
-          groupedPartitions.putIfAbsent(ssi, Lists.newArrayList());
-        }
-        groupedPartitions.get(ssi).add(entry.getKey());
+        groupedPartitions.putIfAbsent(ssi, Lists.newArrayList());
```
Comment (xianjingfeng) [inline client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java:525]: `Lists.newArrayList()` will be invoked everytime.

Comment (zuston) [reply client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java:525]: Emm.. Yes, I forget this problem for `putIfAbsent`

Comment (zuston) [reply client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java:525]: Actually, the original logic is OK for me. Let me recover this part.

**Review by xianjingfeng — APPROVED** [review]
LGTM

**Pull request merged**